### ISSUE TITLE
Redundancy in OMS-elements list: type vs. choice statement

### ISSUE THREAD

**Issue by jktjkt** [ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang#21]
Within the `OMS-elements` list, the model supports multiple "types" or "classes" of elements -- amplifiers, fiber spans and attenuators, and possible ROADMs in future. The distinction on "what type of an element" is currently modeled via two means:

- the `type` leaf which is an `identityref` to a `type-element` identity,
- the `choice` YANG statement and its inner container.

This is redundant and possibly confusing, as shown by the currently commented-out stnzas such as `when "type = 'Edfa'";`.

I think that the `choice` and `case` YANG constructs are a sufficient tool here. They guarantee that exactly one of the possible options is selected. I am not aware of any issues or a possible loss of functionality. Esther pointed out that the leafs contained immediately within a `case` must be dosjoint sets, but that's not a problem because there's an intermediate container (`amplifier`, `fiber`, etc).

**Comment by agva123** [comment]
IMHO there are two reasons this "type" and "choice" statement both exist and are not duplicates:
1) to state the node type explicitly, so client can query a single parameter to know the node type and take appropriate actions. Having just node-type-specific parameters makes the inference of node type implicit.
2) to use YANG to explicitly verify a particular node type against its associated parameters. It is true that only one of the possible options under choice statement can be used at a time, however if a node is type amplifier but wrongfully presents a fiber parameter there will be no verification to discover it. The type will add an extra option of verification.

**Comment by jktjkt** [comment]
> 1. to state the node type explicitly, so client can query a single parameter to know the node type and take appropriate actions. Having just node-type-specific parameters makes the inference of node type implicit.

I think I understand this objection, but I believe that this functionality is still available. Previously, a user had a choice for either comparing the `identityref`, or looking at a presence of a given container. I.e., when checking whether a node is an amplifier, the user could either:

- check `OMS-attributes/OMS-elements[elt-index=...]/type` whether the `identity` that it points to is **derived** from, e.g., `amplifier`,
- **or**, check for existence of the `OMS-attributes/OMS-elements[elt-index=...]/amplifier` data tree node

With patch from PR #22, this is indeed made "implicit", but please note that it is still fully possible to determine a type of any OMS node programmatically. And because this model always uses a container within the `choice` statement, there's no risk of the usual schema vs. data node and "disappearance" of `choice` schema nodes when looking at the data tree.

> 2. to use YANG to explicitly verify a particular node type against its associated parameters. It is true that only one of the possible options under choice statement can be used at a time, however if a node is type amplifier but wrongfully presents a fiber parameter there will be no verification to discover it. The type will add an extra option of verification.

I believe that this is actually guaranteed by the YANG validation. If the data attempted to add a fiber-specific parameter within, say, and amplifier node, that would immediately result in a validation error. As an example, this is how I expect the YANG data to look like with #22 merged:

```xml
<OMS-elements elt-index="0">
  <amplifier>
    <type_variety>foo-bar-123</type_variety>
    <operational>
      <actual-gain>19.5</actual-gain>
      ...
    </operational>
  </amplifier>
</OMS-elements>
<OMS-elements elt-index="1">
  <fiber>
    <type_variety>SSMF</type_variety>
    <length>12345</length>
  </fiber>
</OMS-elements>
```

Do you have an example data showing this problem? What YANG tooling do you use?

**Comment by sergiobelotti** [comment]
Closed by PR #50
